app/src/main/kotlin/org/cru/godtools/ui/dashboard/lessons/LessonsViewModel.kt

```diff
@@ -40,12 +40,21 @@ class LessonsViewModel @Inject constructor(
     private val savedStateHandle: SavedStateHandle,
     val settings: Settings,
 ) : ViewModel() {
-    @OptIn(ExperimentalCoroutinesApi::class)
-    val selectedLanguage = savedStateHandle.getStateFlow(KEY_SAVED_LESSON_LANGUAGE_LOCALE, settings.appLanguage)
-        .flatMapLatest { locale ->
-            languagesRepository.findLanguageFlow(locale).map { it ?: Language(locale) }
+    val appLanguage = settings.appLanguageFlow
+        .distinctUntilChanged { old, new ->
+            if (old != new) {
+                updateSelectedLanguage(Language(new))
+            }
+            old == new
         }
-        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), Language(settings.appLanguage))
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    val selectedLanguage = appLanguage.flatMapLatest {
+        savedStateHandle.getStateFlow(KEY_SAVED_LESSON_LANGUAGE_LOCALE, it)
+            .flatMapLatest { locale ->
+                languagesRepository.findLanguageFlow(locale).map { it ?: Language(locale) }
+            }
+    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), Language(settings.appLanguage))
```

Comment:
this is a very convoluted way to handle this, you should look at using an `init {}` block to monitor the `appLanguageFlow` and trigger `updateSelectedLanguage` from that instead.

Reply:
it's convoluted in the fact that you are triggering a random side effect (`updateSelectedLanguage`) that affects a different flow you are combining later.

In addition, you are re-collecting `savedStateHandle.getStateFlow()` when the actual flow it's returning isn't going to be any different